crates/polars-stream/src/nodes/io_sources/ipc.rs

```diff
@@ -0,0 +1,508 @@
+            },
+        };
+
+        let projection = with_columns
+            .as_ref()
+            .map(|cols| columns_to_projection(cols, &first_metadata.schema))
+            .transpose()?;
+
+        let projected_schema = projection.as_ref().map_or_else(
+            || first_metadata.schema.clone(),
+            |prj| Arc::new(apply_projection(&first_metadata.schema, prj)),
+        );
+
+        Ok(IpcSourceNode {
+            sources,
+
+            row_index,
+            projection,
+            slice,
+
+            rechunk,
+            include_file_paths,
+
+            projected_schema,
+
+            is_finished: AtomicBool::new(false),
+
+            first_metadata,
+        })
+    }
+}
+
+/// Move `slice` forward by `n` and return the slice until then.
+fn slice_take(slice: &mut Range<usize>, n: usize) -> Range<usize> {
+    let offset = slice.start;
+    let length = slice.len();
+
+    assert!(offset < n);
+
+    let chunk_length = (n - offset).min(length);
+
+    let rng = offset..offset + chunk_length;
+
+    *slice = 0..length - chunk_length;
+
+    rng
+}
+
+impl ComputeNode for IpcSourceNode {
+    fn name(&self) -> &str {
+        "ipc_source"
+    }
+
+    fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) -> PolarsResult<()> {
+        assert!(recv.is_empty());
+        assert_eq!(send.len(), 1);
+
+        let is_finished = self.is_finished.load(Ordering::Relaxed);
+        if is_finished {
+            send[0] = PortState::Done;
+        }
+
+        if send[0] != PortState::Done {
+            send[0] = PortState::Ready;
+        }
+
+        Ok(())
+    }
+
+    fn spawn<'env, 's>(
+        &'env mut self,
+        scope: &'s TaskScope<'s, 'env>,
+        recv: &mut [Option<RecvPort<'_>>],
+        send: &mut [Option<SendPort<'_>>],
+        _state: &'s ExecutionState,
+        join_handles: &mut Vec<JoinHandle<PolarsResult<()>>>,
+    ) {
+        assert!(recv.is_empty());
+        assert_eq!(send.len(), 1);
+
+        let senders = send[0].take().unwrap().parallel();
+
+        let slf = &*self;
+
+        join_handles.push(scope.spawn_task(TaskPriority::Low, async move {
+            struct BatchMessage {
+                source_idx: usize,
+                memslice: Arc<MemSlice>,
+                metadata: Arc<FileMetadata>,
+                file_path: Option<Arc<str>>,
+                row_start: IdxSize,
+                slice: Range<usize>,
+                block_range: Range<usize>,
+                seq: u64,
+            }
+
+            let num_senders = senders.len();
+            let (mut distribute, rxs) =
+                distributor_channel::<BatchMessage>(num_senders, DEFAULT_DISTRIBUTOR_BUFFER_SIZE);
+
+            let decoding_tasks = senders
+                .into_iter()
+                .zip(rxs)
+                .map(|(mut send, mut rx)| {
+                    scope.spawn_task(TaskPriority::Low, async move {
```

Comment:
I'd prefer it if we didn't spawn tasks inside tasks unless necessary. When possible I'd like to pre-spawn all tasks during `spawn`.